yarn-project/simulator/src/avm/opcodes/instruction.ts

```diff
@@ -14,14 +15,32 @@ type InstructionConstructor = {
  * It's most important aspects are execute and (de)serialize.
  */
 export abstract class Instruction {
+  /**
+   * Consumes gas and executes the instruction.
+   * This is the main entry point for the instruction.
+   * @param context - The AvmContext in which the instruction executes.
+   */
+  public run(context: AvmContext): Promise<void> {
+    context.machineState.consumeGas(this.gasCost());
+    return this.execute(context);
+  }
+
+  /**
+   * Loads default gas cost for the instruction from the GasCosts table.
+   * Instruction sub-classes can override this if their gas cost is not fixed.
+   */
+  public gasCost(): GasCost {
```

Comment:
Hmm will this work? E.g., if the gas cost depends on the inputs for the operation, you might not have enough info until you read from memory in `execute` (especially if you need to do indirections).

I don't know how fancy gas calculation needs to be, so I'll leave it up to you :)

Reply:
Is there an upper bound we could set for a naive implementation?

In a more sophisticated calculation, could it be two parts? First check that I have enough gas to read from memory, then check that I have enough to perform the op?

Reply:
> Is there an upper bound we could set for a naive implementation?
Depends on how precise you want to be. For ADD, for example, there are 3 operands, all of which could be indirect memory. So you'd have an upper bound for mem access as `2*3`.

Another example is [calldatacopy](https://yp-aztec.netlify.app/docs/public-vm/instruction-set/#isa-section-calldatacopy) where you have mem reads but also a size (known at opcode construction time): there you probably want sth like `2*2 + copySize - cdOffset`.

However, assume that you have a hashing opcode HASH that takes a mem offset for the start of the message to hash, and a mem offset (!) for the size. Then the size is not known at opcode construction time, it will be retrieved from memory (this is to enable hashes in noir over vectors whose size is not fixed at compile time). Then in this case (and probably only in this case, but you should look around) you cannot know how much you'll read.

> In a more sophisticated calculation, could it be two parts?

That separation sounds too add hoc so I'd prefer not. Memory reading is one possible dimension, is there going to be a new one (storage?) and we'll start having 2*dimensions parts? :) If it's going to get too complicated, then maybe it just has to be done in `execute()`.

More generally, should gas cost be possible to know without executing anything at all? Then we cannot have indirect memory addresses. Otherwise if some execution (and revertion) is ok then you could just do the gas calculation as you go.

(I don't have strong opinions on this beyond not having ad-hoc phases)

Reply:
> More generally, should gas cost be possible to know without executing anything at all?

Sounds like probably not then. I think I'd be fine with having the sequencer perform the work to determine the true amount required. Seems like very bounded risk.

> (I don't have strong opinions on this beyond not having ad-hoc phases)

👍